packit/dg_robot.py

```diff
@@ -0,0 +1,256 @@
+    workflow and feeding input should be done up the stack
+
+    The expectation is that we work with a single instance of dist-git, that's the state, and methods of this class
+    interact with the local copy
+    """
+    def __init__(
+            self,
+            github_token: str,
+            pagure_user_token: str,
+            pagure_package_token: str,
+            pagure_fork_token: str,
+            package_name: str,
+            fas_user: str,
+            dist_git_url: str,
+            dist_git_namespace: str = "rpms",
+            dist_git_path: str = None,
+            upstream_project_url: str = ".",  # can be URL or path
+    ) -> None:
+        self.github_token = github_token
+        self.pagure_user_token = pagure_user_token
+        self.pagure_package_token = pagure_package_token
+        self.pagure_fork_token = pagure_fork_token
+
+        self.package_name = package_name
+        self.fas_user = fas_user
+        self.dist_git_url = dist_git_url
+        self.dist_git_namespace = dist_git_namespace
+        self.dist_git_path = dist_git_path
+        self.upstream_project_url = upstream_project_url
+
+        self._distgit = None
+        self._upstream_project = None
+        self._package_config = None
+        self._tmpdir = None
+        self._distgit_spec = None
+        self._upstream_spec = None
+
+    @property
+    def tmpdir(self):
+        if self._tmpdir is None:
+            self._tmpdir = tempfile.mkdtemp(prefix="packit-")
+        return self._tmpdir
+
+    @property
+    def distgit_spec_path(self):
+        return os.path.join(
+            self.distgit.working_dir,
+            f"{self.package_name}.spec"
+        )
+
+    @property
+    def upstream_spec_path(self):
+        return os.path.join(
+            self.upstream.working_dir,
+            f"{self.package_name}.spec"
+        )
+
+    @property
+    def upstream(self):
+        """  """
+        if self._upstream_project is None:
+            self._upstream_project = LocalProject(
+                working_dir=self.upstream_project_url
+            )
+        return self._upstream_project
+
+    @property
+    def distgit_specfile(self):
+        if self._distgit_spec is None:
+            self._distgit_spec = SpecFile(
+                path=self.distgit_spec_path,
+                sources_location=self.distgit.working_dir,
+                changelog_entry=None,
+            )
+        return self._distgit_spec
+
+    @property
+    def upstream_specfile(self):
+        if self._upstream_spec is None:
+            self._upstream_spec = SpecFile(
+                path=self.upstream_spec_path,
+                sources_location=self.upstream.working_dir,
+                changelog_entry=None,
+            )
+        return self._upstream_spec
+
+    @property
+    def distgit(self):
+        """  """
+        if self._distgit is None:
+            working_dir = self.dist_git_path if self.dist_git_path else self.tmpdir
```

Comment:
If I did it correctly, the working directory should be created in temp automatically.

- https://github.com/packit-service/packit/blob/master/packit/local_project.py#L86
- https://github.com/packit-service/packit/blob/master/packit/utils.py#L173

Reply:
`self.dist_git_path` can be passed from CLI

It was super-frustrating for me when it cloned the repo every time I ran packit; especially when the cafe I was in had a flaky wifi. Instead I implemented this: work in existing dist-git repo I have on my laptop.

Reply:
Yes, I agree. Just pass the `self.dist_git_path` to LocalProject and it will use it. It creates the temp-dir only if needed. (Or at least should...;-)

(So, if `self.dist_git_path` is `None`, it creates the temp-dir, otherwise it use the `self.dist_git_path` as working directory.)

Reply:
oki, I will, got it now

Reply:
I am only not sure about the best cleaning behaviour in the `LocalProject`... Now, you should set `working_dir_temporary` to `False` to preserve the tempdir after object removal.